### ISSUE TITLE
Remove Link component

### ISSUE THREAD

**Issue by nekitu** [OurMachinery/themachinery-public#16]
API: Transform and Link are too much for default situations, Transform should be just it, no Link.

**Comment by niklas-ourmachinery** [comment]
I agree that it is a little bit annoying that you have to basically give every entity both a Transform and a Link component and that the Transform component doesn't really do anything in the editor. But I don't like the alternatives that much.

We could merge the features of both the Transform and Link component into a single component that handles both transform and linking. But I don't think that is a good idea:

- It will force entities that just need a Transform to also have all the Link features, which is unnecessarily expensive.
- I like having a more detailed breakdown with small components with single tasks.
- Users may wish to do other kinds of linking than what the Link component offers, for example, constrain the position of an entity based on the position of other entities. If Link and Transform are baked together, this becomes harder.

Another option than merging the two components would be to hide the Transform component from the UI if a link component is present. I'm not super fond of this solution either since it means that there is "invisible" and "magical" stuff happening behind the scene that the user doesn't see. I like being explicit about things.

**Comment by nekitu** [comment]
Link component seems to me superfluous. Entity hierarchy already governs Transform linkage. Its just added granulation that really wont be used by many people. imho, but we'll see, I just did a quick look over the usage and the header file.

**Comment by niklas-ourmachinery** [comment]
Yes, we'll have some more thinking about this. It seems to be a thing that multiple users get confused about, so we should definitely address it somehow (even if it is just in documentation).

However, this are kind of fundamental components of the engine, so it's not something we can change lightly... we have to think hard about what's the right thing to do.

**Comment by nekitu** [comment]
Sure, I just brought it to attention. It is a bit different from traditional ECS, I understand the lego thing, but drastic simplification and abstraction can make the engine's learning even more steeper than it is, plus it might even complicate things in some scenarios. Transform that does what is says it does, is enough to drive the entity's location, and the already in place hierarchy does the rest. But if you have a scenario where Link is absolutely needed, please share, others might see this and understand more.

**Comment by niklas-ourmachinery** [comment]
Well Transform by itself is just a component that holds the entity's transform (pos/rot/scale). It doesn't automatically link its transform to the parent.

In *most* cases you probably want that... but there are cases where you might want to have something as a child entity just for logical/organizing reasons, without linking its position to the parent's position... this is why linking is enabled by explicitly adding a Link component.

**Comment by nekitu** [comment]
Why not a simple bool property in the transform component ? DontUseParentTransform or something

**Comment by niklas-ourmachinery** [comment]
Well, the main issue is if the UI should reflect what is "actually" happening or not.

If we don't care about that we could have a bool property like that. If the property is not checked, we create a Link component "behind the scene" (we still need the Link component to do the actual linking). If it's checked, it will just be the Transform component.

However, I'm vary about doing things "behind the scene", because it means the UI model (which is the users default mental model) no longer matches exactly what is happening when the engine runs. And this can lead to confusion and bugs.

Anyways, I don't think we'll be able to solve it in this thread right now, we'll just have to have a big think about it.

**Comment by nekitu** [comment]
By the amount of text your write, I feel you are thinking about it, which is good :D, you're making the lego bricks too small! :), I have issues since I saw your x[] y[] z[] arrays just to be fast :)), while it is good for speed, it is painful for users, keep a balance in the Force.

**Comment by niklas-ourmachinery** [comment]
Lol! Well hopefully we can do both — great performance and great UI. We’ll see!

**Comment by karl-ourmachinery** [comment]
We should decide on this soon.

To me, this is mostly a question of UI in the property tab. There are clearly nice power-user reasons for keeping the Link and Transform separately, but the current setup is annoying to look at in the property tab. We could keep the Link and Transform component, but change the UI of any Entity so it looks like this (left is before, right is after):

![image](https://user-images.githubusercontent.com/57644882/101234739-252a0c00-36c2-11eb-95ff-2a592787618f.png) ![image](https://user-images.githubusercontent.com/57644882/101234617-3de5f200-36c1-11eb-9e7a-5aee844eeced.png)

The Link Component clearly states that it controls the Transform, so we can avoid the Transform stealing all the attention by being at the top and saying "Controlled by Link Component" like it currently does. If you delete the Link component, then the Transform component is drawn (you sometimes want to delete the Link component and keep the Transform component, but I don't see why you would like to delete the Transform and keep the Link).

I have removed the Scene Tree Node property, I guess we can just skip showing it unless a Scene Tree actually exists. I think users will catch on quickly, and a tiny bit of magic is OK to make these central UIs nicer on the eye.

Also, in this mockup I moved the Add Component button to a toolbar. This is because it sat at and awkward position before. It would have been better at the bottom, but if the list becomes long, then you'd might have to scroll to get to the button. So I figured it would be nice to have in a toolbar at the bottom. Moving it to a toolbar should be easy to do since we should be able to add a toolbar callback to the object being rendered in the properties tab.

All in all I think this would make the entity property UI easier on the eye.
